utils/spark_utils.py

```diff
@@ -759,3 +740,64 @@ def get_standard_concept_ids(spark, concept_ids):
             WHERE ca.concept_id_1 IN ({concept_ids})
         """.format(concept_ids=','.join([str(c) for c in concept_ids])))
     return standard_concept_ids
+
+
+def link_events_with_visits(patient_event):
+    """
+    Links events without a visit_occurrence_id to a same day visit
+    """
+    window = W.partitionBy(['cohort_member_id', 'person_id', 'date'])
+    patient_event = patient_event.withColumn('visit_occurrence_id',
+                                             F.last('visit_occurrence_id', ignorenulls=True).over(window)) \
+        .withColumn('visit_concept_id', F.last('visit_concept_id', ignorenulls=True).over(window))
+    return patient_event
+
+
+def create_visits(patient_event, link_events_visits, gap):
+    """
+    Creates visits as a consecutive sequence of events.
+    The parameter gap controls what events are considered consecutive. A gap of 1
+    means events with 0 days between them are consecutive. A gap of 0 means only same day events
+    are consecutive.
+    """
+    # create date_diff column
+    window = W.partitionBy("cohort_member_id", "person_id").orderBy("date")
+    patient_event = patient_event.withColumn("date_difference",
+                                             F.datediff(patient_event.date,
+                                                        F.lag(patient_event.date, 1).over(
+                                                            window)))
+    # first row per person_id gives nan for date_difference, replace with value 2
+    patient_event = patient_event.na.fill(2, "date_difference")
+
+    if link_events_visits:
+        # only create visits when visit info is missing
+        patient_event = patient_event.withColumn("_flg",
+                                                 F.coalesce(F.when((F.col("date_difference") > gap) & \
+                                                                   (F.col("visit_occurrence_id").isNull()), 1),
+                                                            F.lit(0)))
+
+        # make sure new visit ids are not clashing with existing
+        whole_window = W.partitionBy("cohort_member_id", "person_id")
+        patient_event = patient_event.withColumn("_max", F.max(F.col("visit_occurrence_id")).over(whole_window))
+
+        patient_event = patient_event.withColumn("visit_occurrence_id",
+                                                 F.when(patient_event.visit_occurrence_id.isNull(),
+                                                        F.col("_max") + F.sum("_flg").over(window)) \
+                                                 .otherwise(patient_event.visit_occurrence_id))
```

Comment:
Nitpick: Is the below suggestion equivalent? 
```suggestion
                                                 F. coalesce(F.col("visit_occurrence_id"), F.col("_max") + F.sum("_flg").over(window))
```

Reply:
It is thanks!